synapse/logging/_terse_json.py

```diff
@@ -141,11 +145,49 @@ def TerseJSONToConsoleLogObserver(outFile: IO[str], metadata: dict) -> FileLogOb
 
     def formatEvent(_event: dict) -> str:
         flattened = flatten_event(_event, metadata)
-        return dumps(flattened, ensure_ascii=False, separators=(",", ":")) + "\n"
+        return _encoder.encode(flattened) + "\n"
 
     return FileLogObserver(outFile, formatEvent)
 
 
+@attr.s
+@implementer(IPushProducer)
+class LogProducer(object):
+    """
+    An IPushProducer that writes logs from its buffer to its transport when it
+    is resumed.
+
+    Args:
+        buffer: Log buffer to read logs from.
+        transport: Transport to write to.
+    """
+
+    _buffer = attr.ib(type=deque)
```

Comment:
I wonder whether this becomes simpler (and potentially faster) if we just had one `LogProducer` backed by a `DeferredQueue` (or equivalent)? That way writes just need to insert into the `DeferredQueue` and the `LogProducer` just sits in a loop pulling stuff off the queue?